### PR TITLE
Parameter Acceptor Proxy.

### PR DESCRIPTION
**Pull request by luca-heltai**
This class provides you with the possibility to use existing classes with static member function `declare_parameters` and member function `parse_parameters` with the `ParameterAcceptor` facilities:
~~~
  ParameterAcceptorProxy<Functions::ParsedFunction<2>> f2("Function 2D");
  ParameterAcceptor::initialize("parameters.prm");
~~~
will then work out of the box, without you needing to call `declare_parameters`, and `parse_parameters`.

I'm planning to write a small tutorial program that uses these facilities.

### CODE REVIEW

File: doc/news/changes/minor/20180306LucaHeltai [hunk 1]
```diff
@@ -0,0 +1,6 @@
+New: the ParameterAcceptorProxy class, allows you to wrap a class that provides a static member
```
Comment (jppelteret) [inline doc/news/changes/minor/20180306LucaHeltai]: > The ParameterAcceptorProxy class allows...

File: doc/news/changes/minor/20180306LucaHeltai [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+declare_parameters and a member parse_parameters into a 
+ParameterAcceptor subclass.
```
Comment (jppelteret) [inline doc/news/changes/minor/20180306LucaHeltai]: Please put this onto the previous line

File: include/deal.II/base/parameter_acceptor.h [hunk 3]
```diff
@@ -50,15 +50,19 @@ DEAL_II_NAMESPACE_OPEN
...
- * ParameterAcceptor::parse_parameters() methods, or by calling its
+ * ParameterAcceptor::parse_parameters() methods, by calling its
  * ParameterAcceptor::add_parameter() method for each parameter we want to
- * have. This in turns makes sure that the given parameter is registered in the
- * global parameter handler (by calling ParameterHandler::add_parameter()), at
- * the correct path. If you define all your parameters using the
- * ParameterAcceptor::add_parameter() method, then you don't need to overload
- * any of the virtual methods of this class.
+ * have, or by constructing a ParameterAcceptorProxy class with your own class,
+ * provided that your class implements the declare_parameters and
```
Comment (jppelteret) [inline include/deal.II/base/parameter_acceptor.h]: @p declare_parameters

Comment (jppelteret) [reply include/deal.II/base/parameter_acceptor.h]: @p parse_parameters

File: include/deal.II/base/parameter_acceptor.h [hunk 4]
```diff
@@ -495,6 +498,7 @@ class ParameterAcceptor : public Subscriptor
    */
   void leave_my_subsection(ParameterHandler &prm);
 
+private:
```
Comment (jppelteret) [inline include/deal.II/base/parameter_acceptor.h]: Why is this change necessary? I don't see it called by `ParameterAcceptorProxy`.

File: include/deal.II/base/parameter_acceptor.h [hunk 5]
```diff
@@ -516,6 +520,72 @@ class ParameterAcceptor : public Subscriptor
 
 
 
+/**
+ * A Proxy ParameterAcceptor wrapper for classes that have a static member
```
Comment (jppelteret) [inline include/deal.II/base/parameter_acceptor.h]: > A proxy

File: include/deal.II/base/parameter_acceptor.h [hunk 6]
```diff
@@ -516,6 +520,72 @@ class ParameterAcceptor : public Subscriptor
...
+ * function declare_parameters, and a non virtual parse_parameters method.
```
Comment (jppelteret) [inline include/deal.II/base/parameter_acceptor.h]: >function @p declare_parameters,

> non virtual @p parse_parameters

File: include/deal.II/base/parameter_acceptor.h [hunk 7]
```diff
@@ -516,6 +520,72 @@ class ParameterAcceptor : public Subscriptor
...
+ *
+ * If a class already implements the static member function declare_parameters,
+ * and the method declare_parameters, then it cannot be derived directly from
+ * ParameterAcceptor, as the existing member functions would be incompatible
+ * with the virtual methods with the same name that ParameterAcceptor expects
+ * you to overload. For these cases, you can still use ParameterAcceptor
+ * facilities and logic by wrapping your class inside this proxy. The resulting
+ * class will have the public interface of your SourceClass, and will provide
```
Comment (jppelteret) [inline include/deal.II/base/parameter_acceptor.h]: @p SourceClass

File: include/deal.II/base/parameter_acceptor.h [hunk 8]
```diff
@@ -516,6 +520,72 @@ class ParameterAcceptor : public Subscriptor
...
+ * An example usage is given by the following snippet of code, using
+ * Functions::ParsedFunction as an example source class:
+ *
+ * @code
+ * ParameterAcceptorProxy<Functions::ParsedFunction<2> > fun("Some function");
+ * ParameterAcceptor::initialize("test.prm");
+ * @endcode
+ *
+ * The above snippet of code will initialize ParameterAcceptor::prm with a
+ * section "Some function", and will correctly parse and assing to the object
```
Comment (jppelteret) [inline include/deal.II/base/parameter_acceptor.h]: How does this work if one has an existing parameter file with a single section missing? Does this overwrite the entire file or just append the missing section to it?

File: include/deal.II/base/parameter_acceptor.h [hunk 9]
```diff
@@ -531,6 +601,30 @@ void ParameterAcceptor::add_parameter(const std::string &entry,
...
+  SourceClass(args...),
+  ParameterAcceptor(section_name)
+{};
+
+
+
+template<class SourceClass>
+void  ParameterAcceptorProxy<SourceClass>::declare_parameters(ParameterHandler &prm)
+{
+  return SourceClass::declare_parameters(prm);
```
Comment (jppelteret) [inline include/deal.II/base/parameter_acceptor.h]: `return` is unnecessary here

File: include/deal.II/base/parameter_acceptor.h [hunk 10]
```diff
@@ -531,6 +601,30 @@ void ParameterAcceptor::add_parameter(const std::string &entry,
...
+}
+
+
+
+template<class SourceClass>
+void  ParameterAcceptorProxy<SourceClass>::parse_parameters(ParameterHandler &prm)
+{
+  return SourceClass::parse_parameters(prm);
```
Comment (jppelteret) [inline include/deal.II/base/parameter_acceptor.h]: `return` is unnecessary here too

File: include/deal.II/base/parameter_acceptor.h (hunk 4 above)
Comment (luca-heltai) [reply include/deal.II/base/parameter_acceptor.h]: I have a few other tests (on my personal classes, not in deal.II) that need to know the exact subsection of the given class. Since the `get_section_name` is public, I thought this should be public too, to allow more complex situations (e.g., you need to change the value of a parameter after some conditions are met, and you want to make sure you enter in the right section).

File: include/deal.II/base/parameter_acceptor.h (hunk 8 above)
Comment (luca-heltai) [reply include/deal.II/base/parameter_acceptor.h]: `ParameterAcceptor` just calls `declare_parameters` and `parse_parameters` for each of its subclass. This does not stop you from having other parameters, as long as you declare them first. Nothing is done to the existing ones.

File: include/deal.II/base/parameter_acceptor.h [hunk 11]
```diff
@@ -516,6 +520,72 @@ class ParameterAcceptor : public Subscriptor
 
 
 
+/**
+ * A proxy ParameterAcceptor wrapper for classes that have a static member
+ * function @p declare_parameters, and a non virtual @p parse_parameters method.
+ *
+ * If a class already implements the static member function declare_parameters,
+ * and the method declare_parameters, then it cannot be derived directly from
```
Comment (bangerth) [inline include/deal.II/base/parameter_acceptor.h]: did you mean `parse_parameters()` here?

File: include/deal.II/base/parameter_acceptor.h [hunk 12]
```diff
@@ -516,6 +520,72 @@ class ParameterAcceptor : public Subscriptor
...
+ * ParameterAcceptor, as the existing member functions would be incompatible
+ * with the virtual methods with the same name that ParameterAcceptor expects
+ * you to overload. For these cases, you can still use ParameterAcceptor
```
Comment (bangerth) [inline include/deal.II/base/parameter_acceptor.h]: The statement "if a class...then it cannot be derived directly....". I *think* that the use case you are thinking of here is that someone has already implemented a class and is thinking how to make it compatible with `ParameterAcceptor`. So when you say "cannot be derived" you have in mind a process ("you can't rewrite this class in a way where it then derives from `ParameterAcceptor`") as opposed to a statement about the current way the class is declared ("close considerations will show that the class can not have been declared as derived from `ParameterAcceptor`; we know this even without looking at the actual declarations because if it were, this would result in a conflict and the compiler would reject the class").

Can you maybe rephrase this statement to clarify that you are talking about converting an existing class?

File: include/deal.II/base/parameter_acceptor.h [hunk 13]
```diff
@@ -516,6 +520,72 @@ class ParameterAcceptor : public Subscriptor
...
+ * facilities and logic by wrapping your class inside this proxy. The resulting
+ * class will have the public interface of your @p SourceClass, and will provide
+ * the correct interface to ParameterAcceptor.
+ *
+ * An example usage is given by the following snippet of code, using
+ * Functions::ParsedFunction as an example source class:
+ *
+ * @code
+ * ParameterAcceptorProxy<Functions::ParsedFunction<2> > fun("Some function");
+ * ParameterAcceptor::initialize("test.prm");
```
Comment (bangerth) [inline include/deal.II/base/parameter_acceptor.h]: I forgot -- does the object created in the first step register itself with a singleton instance that you query in the second line?

File: include/deal.II/base/parameter_acceptor.h (hunk 8 above)
Comment (bangerth) [reply include/deal.II/base/parameter_acceptor.h]: assing -> assign

File: include/deal.II/base/parameter_acceptor.h [hunk 14]
```diff
@@ -516,6 +520,72 @@ class ParameterAcceptor : public Subscriptor
...
+ * @endcode
+ *
+ * The above snippet of code will initialize ParameterAcceptor::prm with a
+ * section "Some function", and will correctly parse and assing to the object
+ * `fun` the expression parsed from the file `test.prm`. If non existant, the
```
Comment (bangerth) [inline include/deal.II/base/parameter_acceptor.h]: -> existent

(in fact, maybe "non-existent")

File: include/deal.II/base/parameter_acceptor.h [hunk 15]
```diff
@@ -516,6 +520,72 @@ class ParameterAcceptor : public Subscriptor
...
+   * Default constructor. The argument `section_name` is forwarded to the
+   * constructor of the ParameterAcceptor class, while all other arguments
+   * are passed to the SourceClass constructor.
+   */
+  template<typename... Args>
+  ParameterAcceptorProxy(const std::string section_name, Args... args);
+
+  /**
+   * Overloads the ParameterAcceptor::declare_parameters function. See the base
+   * class for further information.
```
Comment (bangerth) [inline include/deal.II/base/parameter_acceptor.h]: for this function and the next, I suppose that it simply calls the corresponding function in `SourceClass`?

File: include/deal.II/base/parameter_acceptor.h [hunk 16]
```diff
@@ -516,6 +520,72 @@ class ParameterAcceptor : public Subscriptor
...
+ *   set Function constants  =
+ *   set Function expression = 0
+ *   set Variable names      = x,y,t
+ * end
+ * @endcode
+ *
+ * @author Luca Heltai, 2018
+ */
+template<class SourceClass>
+class ParameterAcceptorProxy : public SourceClass, public ParameterAcceptor
```
Comment (bangerth) [inline include/deal.II/base/parameter_acceptor.h]: I think you should state somewhere in the documentation that this class derives from `SourceClass` and that consequently it has the same public interface as `SourceClass`.

File: include/deal.II/base/parameter_acceptor.h [hunk 17]
```diff
@@ -516,6 +520,72 @@ class ParameterAcceptor : public Subscriptor
...
+   * Overloads the ParameterAcceptor::declare_parameters function. See the base
+   * class for further information.
+   */
+  virtual void declare_parameters(ParameterHandler &prm);
+
+  /**
+   * Overloads the ParameterAcceptor::parse_parameters function. See the base
+   * class for further information.
+   */
+  virtual void parse_parameters(ParameterHandler &prm);
```
Comment (bangerth) [inline include/deal.II/base/parameter_acceptor.h]: what happens if the base class has a member function of same name that is also `virtual`?

Comment (luca-heltai) [reply include/deal.II/base/parameter_acceptor.h]: the implementation takes care of this case, because it calls explicitly `SourceClass::parse_parameters`.

File: include/deal.II/base/parameter_acceptor.h (hunk 16 above)
Comment (luca-heltai) [reply include/deal.II/base/parameter_acceptor.h]: Done.

File: include/deal.II/base/parameter_acceptor.h (hunk 17 above)
Comment (bangerth) [reply include/deal.II/base/parameter_acceptor.h]: But the current class then has two base classes, each of which declares its own virtual member function of that name. Which one does the current function overload? I don't think C++ allows this, or does it?

Comment (bangerth) [reply include/deal.II/base/parameter_acceptor.h]: It turns out that one then simply overloads the virtual functions of both base classes. This seems to work:
```
#include <iostream>

class A
{
public:
  virtual void f() = 0;
};

class B
{
public:
  virtual void f() = 0;
};

class C :
  public A, public B
{
public:
  virtual void f() { std::cout << "***\n"; }
};


int main ()
{
  C c;
  ((A*)&c)->f();
  ((B*)&c)->f();
}
```

File: include/deal.II/base/parameter_acceptor.h (hunk 13 above)
Comment (luca-heltai) [reply include/deal.II/base/parameter_acceptor.h]: Yes. ParameterAcceptor has a global `ParameterHandler` member `ParameterAcceptor::prm`, and a global list of classes that register with it. For every class that has been registered, i.e., for every class that was derived by ParameterAcceptor, and that was constructed before the `initialize` function is called, the declare and parse is called with `initialize`.

**Comment by luca-heltai** [comment]
@bangerth  ping?

**Comment by bangerth** [comment]
OK to merge if you squash it a bit.

File: include/deal.II/base/parameter_acceptor.h [hunk 18]
```diff
@@ -516,6 +520,82 @@ class ParameterAcceptor : public Subscriptor
...
+ * # DEAL_II_PACKAGE_VERSION = 9.0.0-pre
+ * subsection Some function
+ *   set Function constants  =
+ *   set Function expression = 0
+ *   set Variable names      = x,y,t
+ * end
+ * @endcode
+ *
+ * The resulting `fun` object, is both a ParsedFunction object and a
+ * ParameterAcceptor one, allowing you to it as a replacement of the
```
Comment (davydden) [inline include/deal.II/base/parameter_acceptor.h]: ..allowing to USE it as...
